crmsh/sh.py

```diff
@@ -0,0 +1,321 @@
+3. AutoShell runs command on cluster nodes. It leverages su, sudo and ssh to obtain a appreciated session on a
+   destination node. It is only available after ssh bootstrap as it depends on the knowledge about cluster node and user
+   configurations.
+
+The LocalShell and SshShell is expected to be used in ssh bootstrap. Once the ssh bootstrap finishes, AuthShell should
+be used.
+"""
+import logging
+import os
+import pwd
+import socket
+import subprocess
+import typing
+
+from . import constants
+from .pyshim import cache
+from . import user_of_host
+from .user_of_host import UserOfHost
+
+logger = logging.getLogger(__name__)
+
+
+class Error(ValueError):
+    def __init__(self, msg, cmd):
+        super().__init__(msg)
+        self.cmd = cmd
+
+
+class AuthorizationError(Error):
+    def __init__(self, cmd: str, host: typing.Optional[str], user: str, msg: str):
+        super().__init__(
+            'Failed to run command on {optional_user}{host}: {msg}: {cmd}'.format(
+                optional_user=f'{user}@' if user is not None else '',
+                host=host, msg=msg, cmd=cmd
+            ),
+            cmd
+        )
+        self.host = host
+        self.user = user
+
+
+class CommandFailure(Error):
+    def __init__(self, cmd: str, host: typing.Optional[str], user: typing.Optional[str], msg: str):
+        if host is None and user is None:
+            super().__init__("Failed to run '{}': {}".format(cmd, msg), cmd)
+        elif user is None:
+            super().__init__("Failed to run command on {}: '{}': {}".format(host, cmd, msg), cmd)
+        elif host is None:
+            super().__init__("Failed to run command as {}: '{}': {}".format(user, cmd, msg), cmd)
+        else:
+            super().__init__("Failed to run command as {}@{}: '{}': {}".format(user, host, cmd, msg), cmd)
+        self.host = host
+        self.user = user
+
+
+class Utils:
+    @staticmethod
+    def decode_str(x: bytes):
+        try:
+            return x.decode('utf-8')
+        except UnicodeDecodeError as e:
+            logger.debug('UTF-8 decode failure', exc_info=e)
+            return x.decode('utf-8', errors='backslashreplace')
+
+
+class LocalShell:
+    """Provides methods to run commands on localhost, both as current user and switching to another user"""
+    @staticmethod
+    @cache
+    def hostname():
+        return socket.gethostname()
+
+    @staticmethod
+    @cache
+    def geteuid() -> int:
+        return os.geteuid()
+
+    @staticmethod
+    @cache
+    def get_effective_user_name() -> str:
+        return pwd.getpwuid(LocalShell.geteuid()).pw_name
+
+    def can_run_as(self, user: str):
+        return self.geteuid() == 0 or self.get_effective_user_name() == user
+
+    def su_subprocess_run(
+            self,
+            user: str,
+            cmd: str,
+            tty=False,
+            preserve_env: typing.Optional[typing.List[str]] = None,
+            **kwargs
+    ):
```

Comment:
Please add a doc string for this method to introduce what is this method for or when is the time to use it.